src/dxvk/dxvk_fence.cpp

```diff
@@ -56,80 +57,94 @@ namespace dxvk {
         Logger::warn(str::format("Importing semaphores of type ", info.sharedType, " not supported by device"));
       }
     }
-
-    m_thread = dxvk::thread([this] { run(); });
   }
 
 
   DxvkFence::~DxvkFence() {
-    m_stop.store(true);
-    m_thread.join();
+    if (m_running.exchange(false, std::memory_order_acq_rel)) {
+      m_condVar.notify_one();
+      m_thread.join();
+    }
 
     m_vkd->vkDestroySemaphore(m_vkd->device(), m_semaphore, nullptr);
   }
 
 
   void DxvkFence::enqueueWait(uint64_t value, DxvkFenceEvent&& event) {
-    std::unique_lock<dxvk::mutex> lock(m_mutex);
+    if (value > getValue()) {
+      {
+        std::unique_lock<dxvk::mutex> lock(m_mutex);
+        m_events.emplace_back(value, std::move(event));
+      }
 
-    if (value > m_lastValue.load())
-      m_queue.emplace(value, std::move(event));
-    else
+      if (!m_running.exchange(true, std::memory_order_acq_rel)) {
```

Comment:
we have to lock anyway so i don't really see any benefit to doing any of this outside the lock, especially since atomic exchange is more expensive than a simple `if (m_running)`.